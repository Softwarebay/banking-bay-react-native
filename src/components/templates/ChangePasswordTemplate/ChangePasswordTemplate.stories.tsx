import { action } from '@storybook/addon-actions';
import { text } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react-native';
import { ViewStory } from 'utils/storybook/ViewStory';

import { ChangePasswordTemplate } from './ChangePasswordTemplate';

storiesOf('Templates/ChangePasswordTemplate', module)
	.addDecorator((getStory) => <ViewStory fullscreen>{getStory()}</ViewStory>)
	.add('Default', () => (
		<ChangePasswordTemplate
			confirmPasswordPlaceholder={text(
				'confirmPasswordPlaceholder',
				'Confirm password'
			)}
			headerText={text('headerText', 'Template header')}
			passwordPlaceholder={text('passwordPlaceholder', 'Password placeholder')}
			onChangePassword={action('onChangePassword')}
		/>
	));
