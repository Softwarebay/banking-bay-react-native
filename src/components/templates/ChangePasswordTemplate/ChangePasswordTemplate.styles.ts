import { StyleSheet } from 'react-native';
import { colors } from 'theme';

export const styles = StyleSheet.create({
	button: {},
	container: {
		backgroundColor: colors.screenBackground,
		flex: 1,
		paddingBottom: 30,
		paddingHorizontal: 20,
		paddingTop: 30,
	},
	headerText: {
		marginBottom: 20,
	},
	input: {
		marginBottom: 15,
	},
});
