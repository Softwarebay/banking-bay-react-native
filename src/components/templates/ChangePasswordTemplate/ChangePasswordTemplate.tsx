import { useRef } from 'react';
import { StatusBar } from 'expo-status-bar';
import { Keyboard, TouchableWithoutFeedback, View } from 'react-native';

import { colors } from 'theme';
import { Button, InputFieldHandle, Text } from 'components/atoms';

import { PasswordInput } from 'components/molecules';
import { styles } from './ChangePasswordTemplate.styles';

export interface Props {
	passwordPlaceholder: string;
	confirmPasswordPlaceholder: string;
	headerText: string;
	onChangePassword: (newPassword: string) => void;
}

export const ChangePasswordTemplate = ({
	headerText,
	passwordPlaceholder,
	confirmPasswordPlaceholder,
	onChangePassword,
}: Props) => {
	const passwordRef = useRef<InputFieldHandle>(null);
	const confirmPasswordRef = useRef<InputFieldHandle>(null);

	const onPress = () => {
		const password = passwordRef.current?.getValue();
		const confirmPassword = confirmPasswordRef.current?.getValue();

		if (password && password === confirmPassword) {
			onChangePassword(password);
		}
	};
	return (
		<>
			<StatusBar backgroundColor={colors.screenBackground} style="dark" />
			<TouchableWithoutFeedback onPress={Keyboard.dismiss}>
				<View style={styles.container}>
					<Text
						style={styles.headerText}
						testID="change-password-template-header-test-id"
						text={headerText}
					/>
					<PasswordInput
						ref={passwordRef}
						placeholder={passwordPlaceholder}
						style={styles.input}
						testID="change-password-template-password-input-test-id"
					/>
					<PasswordInput
						ref={confirmPasswordRef}
						placeholder={confirmPasswordPlaceholder}
						style={styles.input}
						testID="change-password-template-confirm-password-input-test-id"
					/>
					<Button
						style={styles.button}
						testID="change-password-template-button-test-id"
						title="Change Password"
						onPress={onPress}
					/>
				</View>
			</TouchableWithoutFeedback>
		</>
	);
};
