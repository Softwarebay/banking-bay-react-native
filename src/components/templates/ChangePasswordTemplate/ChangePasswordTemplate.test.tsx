import { fireEvent, render, screen } from '@testing-library/react-native';
import testRenderer from 'react-test-renderer';

import { ChangePasswordTemplate } from './ChangePasswordTemplate';

import type { Props } from './ChangePasswordTemplate';

describe('ChangePasswordTemplate', () => {
	const onChangePassword = jest.fn<
		Props['confirmPasswordPlaceholder'],
		[string]
	>();

	const props: Props = {
		headerText: 'Template header',
		passwordPlaceholder: 'Password placeholder',
		confirmPasswordPlaceholder: 'Confirm password placeholder',
		onChangePassword,
	};

	it('should match snapshot', () => {
		const tree = testRenderer
			.create(<ChangePasswordTemplate {...props} />)
			.toJSON();

		expect(tree).toMatchSnapshot();
	});

	it('should render template header', () => {
		render(<ChangePasswordTemplate {...props} />);

		const headerText = screen.getByTestId(
			'change-password-template-header-test-id'
		);

		expect(headerText).toHaveTextContent('Template header');
	});

	it('should render two password inputs with specified placeholders', () => {
		render(<ChangePasswordTemplate {...props} />);

		const inputs = screen.getAllByTestId(
			/^change-password-template-(\w(-\w)?)+-input-test-id/g
		);

		expect(inputs).toHaveLength(2);

		const enterPasswordInputField = screen.getByTestId(
			'input-field-change-password-template-password-input-test-id'
		);

		const confirmPasswordInputField = screen.getByTestId(
			'input-field-change-password-template-confirm-password-input-test-id'
		);

		expect(enterPasswordInputField.props).toHaveProperty(
			'placeholder',
			'Password placeholder'
		);

		expect(confirmPasswordInputField.props).toHaveProperty(
			'placeholder',
			'Confirm password placeholder'
		);
	});

	it('should render initially empty inputs', () => {
		render(<ChangePasswordTemplate {...props} />);

		const enterPasswordInputField = screen.getByTestId(
			'input-field-change-password-template-password-input-test-id'
		);

		const confirmPasswordInputField = screen.getByTestId(
			'input-field-change-password-template-confirm-password-input-test-id'
		);

		expect(enterPasswordInputField.props).toHaveProperty('value', '');
		expect(confirmPasswordInputField.props).toHaveProperty('value', '');
	});

	it("should render 'Change Password' button", () => {
		render(<ChangePasswordTemplate {...props} />);

		const btn = screen.getByTestId('change-password-template-button-test-id');

		expect(btn).toHaveTextContent('Change Password');
	});

	it('should submit form when both inputs have the same value', () => {
		render(<ChangePasswordTemplate {...props} />);

		const enterPasswordInputField = screen.getByTestId(
			'input-field-change-password-template-password-input-test-id'
		);

		const confirmPasswordInputField = screen.getByTestId(
			'input-field-change-password-template-confirm-password-input-test-id'
		);

		fireEvent.changeText(enterPasswordInputField, 'password');
		fireEvent.changeText(confirmPasswordInputField, 'password');

		const btn = screen.getByTestId('change-password-template-button-test-id');

		fireEvent.press(btn);

		expect(onChangePassword).toHaveBeenCalledTimes(1);
		expect(onChangePassword).toHaveBeenCalledWith('password');
	});

	it('should not submit form when inputs have different value', () => {
		render(<ChangePasswordTemplate {...props} />);

		const enterPasswordInputField = screen.getByTestId(
			'input-field-change-password-template-password-input-test-id'
		);

		const confirmPasswordInputField = screen.getByTestId(
			'input-field-change-password-template-confirm-password-input-test-id'
		);

		fireEvent.changeText(enterPasswordInputField, 'password');
		fireEvent.changeText(confirmPasswordInputField, 'drowssap');

		const btn = screen.getByTestId('change-password-template-button-test-id');

		fireEvent.press(btn);

		expect(onChangePassword).not.toHaveBeenCalled();
	});

	it('should not submit form when one or more inputs are empty', () => {
		render(<ChangePasswordTemplate {...props} />);

		const enterPasswordInputField = screen.getByTestId(
			'input-field-change-password-template-password-input-test-id'
		);

		const confirmPasswordInputField = screen.getByTestId(
			'input-field-change-password-template-confirm-password-input-test-id'
		);

		expect(enterPasswordInputField.props).toHaveProperty('value', '');
		expect(confirmPasswordInputField.props).toHaveProperty('value', '');

		const btn = screen.getByTestId('change-password-template-button-test-id');

		fireEvent.press(btn);

		expect(onChangePassword).not.toHaveBeenCalled();
	});
});
