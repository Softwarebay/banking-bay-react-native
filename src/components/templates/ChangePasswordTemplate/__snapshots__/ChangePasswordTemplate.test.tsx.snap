// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChangePasswordTemplate should match snapshot 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "#FAFCFF",
      "flex": 1,
      "paddingBottom": 30,
      "paddingHorizontal": 20,
      "paddingTop": 30,
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#4C4C60",
          "fontFamily": "MulishRegular",
          "fontSize": 17,
          "lineHeight": 27.2,
        },
        Object {
          "marginBottom": 20,
        },
      ]
    }
    testID="change-password-template-header-test-id"
  >
    Template header
  </Text>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
        },
        Object {
          "flexDirection": "row",
        },
        Array [
          Object {
            "backgroundColor": "#FFFFFF",
            "borderRadius": 10,
            "shadowColor": "#161966",
            "shadowOffset": Object {
              "height": 6,
              "width": 0,
            },
            "shadowOpacity": 0.08,
            "shadowRadius": 16,
          },
          Array [
            Object {
              "height": 55,
              "paddingHorizontal": 20,
              "paddingVertical": 14,
            },
            Object {
              "marginBottom": 15,
            },
          ],
        ],
      ]
    }
    testID="change-password-template-password-input-test-id"
  >
    <TextInput
      autoCorrect={false}
      keyboardType="default"
      onChangeText={[Function]}
      placeholder="Password placeholder"
      secureTextEntry={true}
      style={
        Array [
          Object {
            "color": "#1B1D4D",
            "fontFamily": "MulishRegular",
            "fontSize": 15,
          },
          Object {
            "flex": 1,
            "marginRight": 20,
          },
        ]
      }
      testID="input-field-change-password-template-password-input-test-id"
      value=""
    />
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      hitSlop={20}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "alignItems": "center",
            "justifyContent": "center",
          },
          Object {
            "right": 0,
          },
          false,
        ]
      }
      testID="icon-btn-change-password-template-password-input-test-id"
    >
      <SvgMock
        color="#1B1D4D"
        height={16}
        testID="eye-off-icon-test-id"
        width={16}
      />
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
        },
        Object {
          "flexDirection": "row",
        },
        Array [
          Object {
            "backgroundColor": "#FFFFFF",
            "borderRadius": 10,
            "shadowColor": "#161966",
            "shadowOffset": Object {
              "height": 6,
              "width": 0,
            },
            "shadowOpacity": 0.08,
            "shadowRadius": 16,
          },
          Array [
            Object {
              "height": 55,
              "paddingHorizontal": 20,
              "paddingVertical": 14,
            },
            Object {
              "marginBottom": 15,
            },
          ],
        ],
      ]
    }
    testID="change-password-template-confirm-password-input-test-id"
  >
    <TextInput
      autoCorrect={false}
      keyboardType="default"
      onChangeText={[Function]}
      placeholder="Confirm password placeholder"
      secureTextEntry={true}
      style={
        Array [
          Object {
            "color": "#1B1D4D",
            "fontFamily": "MulishRegular",
            "fontSize": 15,
          },
          Object {
            "flex": 1,
            "marginRight": 20,
          },
        ]
      }
      testID="input-field-change-password-template-confirm-password-input-test-id"
      value=""
    />
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      hitSlop={20}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "alignItems": "center",
            "justifyContent": "center",
          },
          Object {
            "right": 0,
          },
          false,
        ]
      }
      testID="icon-btn-change-password-template-confirm-password-input-test-id"
    >
      <SvgMock
        color="#1B1D4D"
        height={16}
        testID="eye-off-icon-test-id"
        width={16}
      />
    </View>
  </View>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        false,
        Object {},
      ]
    }
    testID="change-password-template-button-test-id"
  >
    <ViewManagerAdapter_ExpoLinearGradient
      proxiedProperties={
        Object {
          "colors": Array [
            4288142079,
            4279654301,
          ],
          "endPoint": Array [
            1,
            1,
          ],
          "locations": undefined,
          "startPoint": Array [
            0,
            0,
          ],
        }
      }
      style={
        Object {
          "alignItems": "center",
          "borderRadius": 10,
          "height": 55,
          "justifyContent": "center",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#FFFFFF",
              "fontFamily": "MulishSemiBold",
              "fontSize": 17,
            },
            Object {
              "color": "#FFFFFF",
            },
          ]
        }
      >
        Change Password
      </Text>
    </ViewManagerAdapter_ExpoLinearGradient>
  </View>
</View>
`;
